src/golang/lib/workflow/dag/workflow_dag.go

```diff
@@ -345,6 +354,21 @@ func (w *workflowDagImpl) OperatorInputs(op operator.Operator) ([]artifact.Artif
 	return artifacts, nil
 }
 
+func (w *workflowDagImpl) OperatorParents(op operator.Operator) ([]operator.Operator, error) {
+	inputArtifactIds, ok := w.opToInputArtifacts[op.ID()]
+	if !ok {
+		return nil, errors.Newf("Unable to find operator %s (%s) on dag.", op.ID(), op.Name())
+	}
+
+	// Since each artifact can only be generated by one operator, the size of the input operators
+	// and the size of the input artifacts are the same.
+	ops := make([]operator.Operator, 0, len(inputArtifactIds))
+	for _, artifactId := range inputArtifactIds {
+		ops = append(ops, w.operators[w.artifactToInputOp[artifactId]])
```

Comment:
Can we assume that the map lookups will always succeed or should we add error handling?

Reply:
Yep we should add some error handling here, will do.